### PR TITLE
JDK-8293978: Duplicate simple loop back-edge will crash the vm

### PR DESCRIPTION
**Pull request by casparcwang**
Duplicate back-edge of the following simple loop will make jvm crash.

image

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293978](https://bugs.openjdk.org/browse/JDK-8293978): Duplicate simple loop back-edge will crash the vm

### Reviewers
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [474bbeba](https://git.openjdk.org/jdk/pull/10329/files/474bbeba4b7f005c79f8ede0bcc03838b0dd38a7)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [b9f86cde](https://git.openjdk.org/jdk/pull/10329/files/b9f86cde8aa0f4e39edbcd34858a59b7c3cb956e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10329/head:pull/10329` \
`$ git checkout pull/10329`

Update a local copy of the PR: \
`$ git checkout pull/10329` \
`$ git pull https://git.openjdk.org/jdk pull/10329/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10329`

View PR using the GUI difftool: \
`$ git pr show -t 10329`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10329.diff">https://git.openjdk.org/jdk/pull/10329.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopopts.cpp [hunk 1]
```diff
@@ -3969,7 +3969,7 @@ bool PhaseIdealLoop::duplicate_loop_backedge(IdealLoopTree *loop, Node_List &old
     exit_test = back_control->in(0)->as_If();
   }
 
-  if (idom(region)->is_Catch()) {
+  if (idom(region)->is_Catch() || region == head) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/loopopts.cpp:3972]: That check could be done earlier I think at the test for:
 `if (!incr->is_Phi()) {`

**Review by rwestrel — CHANGES_REQUESTED** [review]

File: src/hotspot/share/opto/loopopts.cpp (hunk 1 above)
Comment (casparcwang) [reply src/hotspot/share/opto/loopopts.cpp]: Thank your for the suggestion! I have move the check to `if (!incr->is_Phi())`

**Review by rwestrel — APPROVED** [review]
Looks good to me.

File: test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  * @summary Crash in PhaseIdealLoop::verify_strip_mined_scheduling
+  *
+  * @run main TestDuplicateSimpleLoopBackedge
+  *
+  */
+
+import jdk.test.lib.Utils;
+
+class Foo {
+    static void c(Byte[] a, Byte d) {
```
[the remark this answers is not in the document]
Comment (casparcwang) [reply test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java]: done

File: test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        for (int e = 0; e < a.length; e++)
+            a[e] = 0;
+    }
+}
+
+public class TestDuplicateSimpleLoopBackedge {
```
[the remark this answers is not in the document]
Comment (casparcwang) [reply test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java]: done

File: test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                    for (int i = 0; i < 10000; ++i) {
+                        n.j(args);
+                    }
+                }
+            };
+        // Give thread some time to trigger compilation
+        thread.setDaemon(true);
+        thread.start();
+        Thread.sleep(Utils.adjustTimeout(4000));
+    }
```
[the remark this answers is not in the document]
Comment (casparcwang) [reply test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java]: Thank you for the review, the test has been changed according to the suggestions.

File: test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  *
+  * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+  * or visit www.oracle.com if you need additional information or have any
+  * questions.
+  */
+
+ /**
+  * @test
+  * @bug 8293978
+  * @library /test/lib
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java:27]: Not required anymore.

File: test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  * @summary Crash in PhaseIdealLoop::verify_strip_mined_scheduling
+  *
+  * @run main/othervm -Xbatch TestDuplicateSimpleLoopBackedge
+  *
+  */
+
+public class TestDuplicateSimpleLoopBackedge {
+    static void zero(Byte[] a) {
+        for (int e = 0; e < a.length; e++)
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java:36]: I think it's better to always add braces, same below for the `while` and `if`.

**Review by chhagedorn — APPROVED** [review]
Otherwise, looks good!

File: test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java (hunk 6 above)
Comment (casparcwang) [reply test/hotspot/jtreg/compiler/c2/TestDuplicateSimpleLoopBackedge.java]: done

**Pull request closed without merging**